File: subsetsum.js
function max_signif(arr) {
    for (let i = 0; i <= 64; i++) {
        let roundedEqs = arr.map(a => a.toFixed(i) == a).reduce(
            (a, b) => a && b);
        if (roundedEqs) {
            return i;
        }
    }
    throw new Error('Digits are too precise');
}

function subset_resistor(res_vals, tgt) {
    let n = res_vals.length
    let X = max_signif(res_vals) + 1;

    let dp = []
    for (let i = 0; i <= n; i++) {
        dp.push([]);
        for (let j = 0; j <= n; j++) {
            dp[dp.length - 1].push(new Set());
        }
    }

    let convertResistance = r => parseFloat(r.toFixed(X));

    dp[0][0].add(0)

    for (let l = 0; l <= n; l++) {
        if (l >= 1) {
            for (let i = 0; i <= n; i++) {
                dp[i][l] = new Set([...dp[i][l-1]]);
            }
        }
        if (l % 2 == 0) {
            for (let i = 1; i <= n; i++) {
                let weight = res_vals[i-1];
                for (let j of dp[i-1][l]) {
                    dp[i][l].add(convertResistance(j));
                    dp[i][l].add(convertResistance(j + weight));
                }
            }
        } else {
            for (let i = 1; i <= n; i++) {
                let weight = res_vals[i-1];
                for (let j of dp[i-1][l]) {
                    s = 1/(1/j + 1/weight);
                    dp[i][l].add(convertResistance(j));
                    dp[i][l].add(convertResistance(s));
                }
            }
        }
    }
    result = dp[n][n].has(tgt);

    if (result) {
        let soln = [];
        let j = tgt;
        let l = n;
        for (i = n; i >= 1; i--) {
            if (j == 0) {
                break;
            }

            let l_ = l;
            while (l > 0 && dp[i][l].has(j) && dp[i][l-1].has(j)) {
                l = l - 1;
            }
            if (l != l_) {
                if (l % 2 == 0) {
                    soln = ["+", ...soln];
                } else {
                    soln = ["||", ...soln];
                }
            }

            let weight = res_vals[i-1];
            if (dp[i][l].has(j) && !(dp[i-1][l].has(j))) {
                soln = [weight, ...soln]
                if (l % 2 == 0) {
                    j = convertResistance(j-weight);
                } else {
                    j = convertResistance(1/(1/j - 1/weight))
                }
            }
            // console.log(`i = ${i}, l = ${l}, j = ${j}`);
        }
        return { result: true, soln: soln };
    } else {
        return { result: false, soln: [] };
    }
}

File: subsetsum.jl
module SubsetSum

export subsetsum, subset_resistor

function maxsignif(A::Array{T}) where {T <: AbstractFloat}
    for X in 0:64
        rounded = [round(i, digits=X) for i in A]
        if rounded == A
            return X
        end
    end
    raise(ArgumentError, "way too precise!")
end

function subsetsum(A::Array{T, 1}, tgt::T1) where {T <: Real, T1 <: Real}
    n = length(A)
    dp = [Set{Real}() for _ in 1:n+1]
    X = maxsignif(A) + 1
    # @show X
    for k in 1:n+1
        push!(dp[k], 0)
    end
    for i in 2:n+1
        weight = A[i-1]
        push!(dp[i], round(weight, digits=X))
        for j in dp[i-1]
            push!(dp[i], round(j, digits=X))
            push!(dp[i], round(j+weight, digits=X))
        end
    end
    result = in(tgt, dp[n+1])
    soln = []
    # @show dp
    if result
        j = tgt
        for i′ in 0:n-1
            if j == 0
                break
            end
            i = n + 1 - i′ # i from n+1 to 2
            weight = A[i-1]
            if in(j, dp[i]) && !in(j, dp[i-1])
                prepend!(soln, weight)
                j = round(j - weight, digits=X)
            end
        end
    end
    return (result, soln)
end

function subset_resistor(A::Array{T, 1}, tgt::T1) where {T <: Real, T1 <: Real}
    n = length(A)
    X = maxsignif(A) + 1

    J = max(findmax(A)[1], tgt)

    dp = [Set() for _ in zeros(n+1, n+1)]
    # dp = zeros(n+1, J+1, n+1)
    push!(dp[1,1], 0)
    # dp[:, 1, 1] .= 1
    for l in 1:n+1
        if l >= 2
            for i in 1:n+1
                union!(dp[i, l], dp[i, l-1])
            end
        end
        if mod(l, 2) == 1
            for i in 2:n+1
                weight = A[i-1]
                for j in dp[i-1, l]
                    push!(dp[i, l], round(j, digits=X))
                    push!(dp[i, l], round(j+weight, digits=X))
                end
            end
        else
            for i in 2:n+1
                weight = A[i-1]
                for j in dp[i-1, l]
                    s = 1/(1/j + 1/weight)
                    push!(dp[i, l], round(j, digits=X))
                    push!(dp[i, l], round(s, digits=X))
                end
            end
        end
    end


    result = in(tgt, dp[n+1, n+1])

    if result
        soln = ""
        j = tgt
        l = n+1
        for i′ in 0:n
            i = n+1 - i′ # i from n+1 to 2
            # if in(j, dp[i, l]) && !in(j, dp[i, l-1])  break end
            #
            if j == 0
                soln = "0.0 $soln"
                break
            end
            l′ = l
            while l > 1 && in(j, dp[i, l]) && in(j, dp[i, l-1])
                l = l - 1
            end
            if l == l′
                soln = soln
            elseif mod(l, 2) == 1
                soln = "+ $soln"
            else
                soln = "|| $soln"
            end

            weight = A[i-1]
            if in(j, dp[i, l]) && !in(j, dp[i-1, l])
                s1 = soln
                soln = "$weight $soln"
                if s1 == ""
                    soln = "$weight"
                end
                if mod(l, 2) == 1
                    j = round(j-weight, digits=X)
                else 
                    j = round(1/(1/j - 1/weight), digits=X)
                end
            end
        end
        return (true, soln)
    else
        return (false, [])
    end
end


end
